File: julia/src/benchmarks/LogParser.jl
mutable struct LogParser <: AbstractBenchmark
    lines_count::Int64
    log::String
    checksum_val::UInt32

    function LogParser()
        lines_count = Helper.config_i64("Etc::LogParser", "lines_count")
        new(lines_count, "", UInt32(0))
    end
end

name(b::LogParser)::String = "Etc::LogParser"

const PATTERN_NAMES = [
    "errors",
    "bots",
    "suspicious",
    "ips",
    "api_calls",
    "post_requests",
    "auth_attempts",
    "methods",
    "emails",
    "passwords",
    "tokens",
    "sessions",
    "peak_hours",
]

const PATTERNS = [
    r" [5][0-9]{2} | [4][0-9]{2} ",
    Regex("bot|crawler|scanner|spider|indexing|crawl|robot|spider", "i"),
    Regex("etc/passwd|wp-admin|\\.\\./", "i"),
    r"\d+\.\d+\.\d+\.35",
    r"/api/[^ \" ]+",
    r"POST [^ ]* HTTP",
    Regex("/login|/signin", "i"),
    Regex("get|post|put", "i"),
    r"[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}",
    r"password=[^&\s\"]+",
    r"token=[^&\s\"]+|api[_-]?key=[^&\s\"]+",
    r"session[_-]?id=[^&\s\"]+",
    r"\[\d+/\w+/\d+:1[3-7]:\d+:\d+ [+\-]\d+\]",
]

const IPS = ["192.168.1.$i" for i = 1:255]
const METHODS = ["GET", "POST", "PUT", "DELETE"]
const PATHS = [
    "/index.html",
    "/api/users",
    "/admin",
    "/images/logo.png",
    "/etc/passwd",
    "/wp-admin/setup.php",
]
const STATUSES = [200, 201, 301, 302, 400, 401, 403, 404, 500, 502, 503]
const AGENTS = ["Mozilla/5.0", "Googlebot/2.1", "curl/7.68.0", "scanner/2.0"]
const USERS = ["john", "jane", "alex", "sarah", "mike", "anna", "david", "elena"]
const DOMAINS =
    ["example.com", "gmail.com", "yahoo.com", "hotmail.com", "company.org", "mail.ru"]

function generate_log_line(i::Int)::String
    idx = i - 1
    result = IOBuffer()

    write(result, IPS[idx%length(IPS)+1])
    write(
        result,
        " - - [",
        string(idx % 31 + 1),
        "/Oct/2023:",
        string(idx % 60),
        ":55:36 +0000] \"",
    )
    write(result, METHODS[idx%length(METHODS)+1], " ")

    if idx % 3 == 0
        write(
            result,
            "/login?email=",
            USERS[idx%length(USERS)+1],
            string(idx % 100),
            "@",
            DOMAINS[idx%length(DOMAINS)+1],
            "&password=secret",
            string(idx % 10000),
        )
    elseif idx % 5 == 0
        write(result, "/api/data?token=")
        for _ = 1:((idx%3)+1)
            write(result, "abcdef123456")
        end
    elseif idx % 7 == 0
        write(result, "/user/profile?session_id=sess_", string(idx * 12345, base = 16))
    else
        write(result, PATHS[idx%length(PATHS)+1])
    end

    write(
        result,
        " HTTP/1.1\" ",
        string(STATUSES[idx%length(STATUSES)+1]),
        " 2326 \"http://",
        DOMAINS[idx%length(DOMAINS)+1],
        "\" \"",
        AGENTS[idx%length(AGENTS)+1],
        "\"\n",
    )

    return String(take!(result))
end

function prepare(b::LogParser)
    log_builder = IOBuffer()
    for i = 1:b.lines_count
        write(log_builder, generate_log_line(i))
    end
    b.log = String(take!(log_builder))
end

function run(b::LogParser, iteration_id::Int64)
    matches = Dict{String,Int}()

    for (idx, pattern) in enumerate(PATTERNS)
        name = PATTERN_NAMES[idx]
        count = length(collect(eachmatch(pattern, b.log)))
        matches[name] = count
    end

    total = 0
    for v in values(matches)
        total += v
    end

    b.checksum_val += UInt32(total)
end

function checksum(b::LogParser)::UInt32
    return b.checksum_val
end

File: nim/src/benchmarks/logparser.nim
import std/[strutils, strformat, sequtils, algorithm, math, tables, re]
import ../benchmark
import ../helper

type
  LogParser* = ref object of Benchmark
    linesCount: int
    log: string
    checksumVal: uint32
    patterns: seq[tuple[name: string, pattern: Regex]]

const
  IPS = block:
    var ips: seq[string]
    for i in 1..255:
      ips.add("192.168.1." & $i)
    ips

  METHODS = ["GET", "POST", "PUT", "DELETE"]
  PATHS = [
    "/index.html", "/api/users", "/admin",
    "/images/logo.png", "/etc/passwd", "/wp-admin/setup.php"
  ]
  STATUSES = [200, 201, 301, 302, 400, 401, 403, 404, 500, 502, 503]
  AGENTS = ["Mozilla/5.0", "Googlebot/2.1", "curl/7.68.0", "scanner/2.0"]
  USERS = ["john", "jane", "alex", "sarah", "mike", "anna", "david", "elena"]
  DOMAINS = ["example.com", "gmail.com", "yahoo.com", "hotmail.com",
      "company.org", "mail.ru"]

  PATTERN_NAMES = [
    "errors", "bots", "suspicious", "ips", "api_calls", "post_requests",
    "auth_attempts", "methods", "emails", "passwords", "tokens", "sessions", "peak_hours"
  ]

  PATTERN_STRS = [
    " [5][0-9]{2} | [4][0-9]{2} ",
    "(?i)bot|crawler|scanner|spider|indexing|crawl|robot|spider",
    "(?i)etc/passwd|wp-admin|\\.\\./",
    "\\d+\\.\\d+\\.\\d+\\.35",
    "/api/[^ \" ]+",
    "POST [^ ]* HTTP",
    "(?i)/login|/signin",
    "(?i)get|post|put",
    "[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,}",
    "password=[^&\\s\"]+",
    "token=[^&\\s\"]+|api[_-]?key=[^&\\s\"]+",
    "session[_-]?id=[^&\\s\"]+",
    "\\[\\d+/\\w+/\\d+:1[3-7]:\\d+:\\d+ [+\\-]\\d+\\]",
  ]

proc generateLogLine(i: int): string =
  result = newStringOfCap(200)
  result.add IPS[i mod IPS.len]
  result.add &" - - [{i mod 31}/Oct/2023:{i mod 60}:55:36 +0000] \""
  result.add METHODS[i mod METHODS.len]
  result.add " "

  if i mod 3 == 0:
    result.add &"/login?email={USERS[i mod USERS.len]}{i mod 100}@{DOMAINS[i mod DOMAINS.len]}&password=secret{i mod 10000}"
  elif i mod 5 == 0:
    result.add "/api/data?token="
    for _ in 0..(i mod 3):
      result.add "abcdef123456"
  elif i mod 7 == 0:
    result.add &"/user/profile?session_id=sess_{toHex(i * 12345)}"
  else:
    result.add PATHS[i mod PATHS.len]

  result.add &" HTTP/1.1\" {STATUSES[i mod STATUSES.len]} 2326 \"http://{DOMAINS[i mod DOMAINS.len]}\" \"{AGENTS[i mod AGENTS.len]}\"\n"

proc newLogParser(): Benchmark =
  LogParser()

method name(self: LogParser): string = "Etc::LogParser"

method prepare(self: LogParser) =
  self.linesCount = self.config_val("lines_count").int

  var sb = newStringOfCap(self.linesCount * 200)
  for i in 0..<self.linesCount:
    sb.add(generateLogLine(i))

  self.log = sb

  self.patterns = newSeq[tuple[name: string, pattern: Regex]](PATTERN_NAMES.len)
  for i in 0..<PATTERN_NAMES.len:
    self.patterns[i] = (PATTERN_NAMES[i], re(PATTERN_STRS[i]))

  self.checksumVal = 0

method run(self: LogParser, iteration_id: int) =
  var matches = initTable[string, int]()

  for (name, pattern) in self.patterns:
    var count = 0
    var pos = 0
    while pos < self.log.len:
      let (first, last) = self.log.findBounds(pattern, pos)
      if first < 0:
        break
      inc count
      pos = last + 1
    matches[name] = count

  var total = 0
  for count in matches.values:
    total += count
  self.checksumVal += total.uint32

method checksum(self: LogParser): uint32 =
  return self.checksumVal

registerBenchmark("Etc::LogParser", newLogParser)
